---
title: "2018-11-25 20min"
output: html_document
---

---
---
# Identify the subset of the DEGs that are Pho4-dependent
As a transcription factor, Pho4 translocate into the nucleus only under phosphate-starvation. In order to get the genes that potentially regulated by Pho4, we need to compare the subset of DEGs of WT and del4 strains under phosphate-depletion, and find the subset of DEGs that only regulated by WT strains but not by del4 deleted strains in early and late groups.

To call the DEGs of pho4 deleted strains, we have to compare the the differential expression of delpho4 strains under normal or phosphate-depleted conditions. Therefore, our criteria for calling the DEG list is to get genes that differentially expressed in at least one time point treatment with log2(fold change) higher than 2 and adjusted p value (pairwise F-test) less than 0.05. 

## Identify the subset of DEGs in del4 groups
Control group:del4 pre (untreated)
Experimental group: del4 treated (from 20min to 240min)

| Experimental group | Control Group | Comment |
| -------- | :---------- | :--------: | ------- |
| del4 20'  | del4 pre  | 20 min up- and down- regulated genes upon phosphate starvation |
| del4 30'  | del4 pre  | 30 min up- and down- regulated genes upon phosphate starvation |
| del4 45'  | del4 pre  | 45 min up- and down- regulated genes upon phosphate starvation |
| del4 60'  | del4 pre  | 60 min up- and down- regulated genes upon phosphate starvation |
| del4 90'  | del4 pre  | 90 min up- and down- regulated genes upon phosphate starvation |
| del4 120' | del4 pre  | 120 min up- and down- regulated genes upon phosphate starvation |
| del4 150' | del4 pre  | 150 min up- and down- regulated genes upon phosphate starvation |
| del4 180' | del4 pre  | 180 min up- and down- regulated genes upon phosphate starvation |
| del4 240' | del4 pre  | 240 min up- and down- regulated genes upon phosphate starvation | 


---
---

## Prepare data

```{r}
##source("http://www.bioconductor.org/biocLite.R")
##biocLite("VennDiagram")
##biocLite("pheatmap")
##biocLite("NMF")
##biocLite("aroma.light")
##install.packages("statmod")
##install.packages("gtools")
library(gtools)
library(statmod)
library(limma)         ## for differential gene expression analysis
library(edgeR)         ## for DGEList()
require(hexbin)
require(ggplot2)       ## for plotting
require(data.table)    ## for fast importing and manipulating data tables
require(cowplot)       ## for some useful defaults in ggplot2
require(RColorBrewer)  ## for plotting
require(gplots)        ## for heatmap plot
library(NMF)           ## for heatmap plot
library(aroma.light)   ## for exploratory analysis
library(knitr)
library(hexbin)
library(VennDiagram)
library(pheatmap)
library(statmod)
```

# 1. Input raw data

```{r load_data}
raw <- fread("../sandbox/Ex009_reads_per_transcript_2017-10-18.txt", check.names = FALSE, stringsAsFactors = FALSE)

# annotation
anno.file <- fread("../data/annotation/C_glabrata_gene_for_mapping_s02-m07-r04.bed")
names(anno.file) <- c("Chr","Start","End","GeneID","Not.Use","Strand","GeneName","Type.of.Gene") 
anno <- anno.file[,c("Chr","GeneID","GeneName","Type.of.Gene")]
setkey(anno, "GeneID")
cgToSc <- fread("../data/annotation/C_glabrata_CBS138_S_cerevisiae_orthologs.txt")
names(cgToSc) <- c("cgid","cgname","cgid2","scid","scname","scid2")
setkey(cgToSc, "cgid")

## names() sets names for the object 
## setkey() sorts a data.table and marks it as sorted, the sorted columns are the key. The columns are sorted in ascending order always


# filter dataset to remove very lowly expressed genes
# 1. examine the distribution of reads for each gene across all samples, to establish a threshold
S <- rowSums(raw[,1:40])
Lib.size<-colSums(raw[,1:40])
#plot(S)
#plot(density(S))
plot(density(log10(S)))
plot(density(log10(Lib.size)))
sprintf("The number of genes with less than 40 total counts across 40 samples, i.e. 1 read per sample on average is %d, and those with less than 100 total counts, or 2.5 reads per sample, is %d", sum(S<=40), sum(S<=100))
print("Let's try 40 reads as a cutoff for dropping genes with low or no expression")

##rowSums(),rowMeans(),colSums() form row and column sums and means for numeric arrays(or data frames)
##density() computes kernal and bandwidth(sd of kernal), evaluates the density at specific points for data
##sprintf() returns a character vector containing a combination of text and variable values; %d conversion by as.character is used for non-character arguments d:interger value
##print() returns a character vector


# filter dataset
isexpr <- (S <= 40)
use.genes <- grepl("ncRNA|ORF|pseudogene", anno$Type.of.Gene)
use <- (!isexpr & use.genes)
mat <- as.matrix(raw[use, 1:40])
anno.all <- anno; anno <- anno.all[use]

## ??? isexpr <- (S <= 40)
## grepl() searches for mathces to argument pattern with each element, 
   # grepl("1",c(1,2,3)); grepl("ncRNA|ORF|pseudogene",c("ORF|Uncharacterized","pseudogene"))
## sub() abd gsub() perform replacement of the fist and all matches respectively
## ??? (!isexpr&use.genes), ! indicated NOT, & and && indicates AND, | and || indicate OR
## ??? why u can write raw[use,1:40], use is a matrix with logic values


# write out the filtered table (uncomment if changes are made to the filtering steps above)
#write.table(as.data.frame(raw[use]), file = paste("../sandbox/Ex009_reads_per_transcript_filtered_", Sys.Date(), ".txt", sep = ""), quote = FALSE, row.names = FALSE)

## ??? how u get gene.names into this new table
##need to go out and separate one column that cotained all the data to different columns 

```

# 2. coerce columns to do differentially expressed genes (DEG) analysis
## pho4-deleted groups & wt groups
```{r}
rawdata <- fread("../sandbox/Ex009_reads_per_transcript_filtered_2018-12-02.txt", check.names = FALSE, stringsAsFactors = FALSE)
sample <- fread("../data/sample_sheet/Ex009_experiment_set_up_20171019.csv",check.names = FALSE, stringsAsFactors = FALSE)
ind.del4 <- which(sample$Genotype=="del4"&!sample$Timepoint=="del80"&!grepl("prime",sample$Sample))
ind.sdel4 <- sample[ind.del4,]
ind <- match(c("gene.names",ind.sdel4$Sample),names(rawdata))
ind.del4 <-rawdata[,ind,with=F]

ind.wt <- which(sample$Genotype=="wt"&!sample$Timepoint=="del80"&!grepl("prime",sample$Sample))
ind.swt <- sample[ind.wt,]
ind2 <- match(c("gene.names",ind.swt$Sample),names(rawdata))
ind.wt <-rawdata[,ind2,with=F]
```


# 3. normalization
## pho4-deleted groups & wt groups
```{r}
y.del4<-DGEList(counts=ind.del4[,2:18],genes= ind.del4$gene.names)
y.del4<-calcNormFactors(y.del4) 
plotMDS(y.del4)
plotMDS(cpm(y.del4,log=TRUE))
### calculate log transformed, normalized count
cpm.un.d4 <- cpm( y.del4, normalized.lib.sizes = FALSE, log = FALSE)
cpm.lg.d4 <- cpm( y.del4, normalized.lib.sizes = FALSE, log = TRUE) 
cpm.nor.d4 <- cpm(y.del4,normalized.lib.sizes = TRUE,log= FALSE)
cpm.nlg.d4 <- cpm(y.del4,normalized.lib.sizes = TRUE,log= TRUE, prior.count=0.5)
boxplot(cpm.un.d4, outline= F)
boxplot(cpm.lg.d4, col="blue",outline= F)
boxplot(cpm.nor.d4, col="red",outline = F)
boxplot(cpm.nlg.d4, col="green",outline = F)
cpm.lgd4.names <- cbind(gene.names=ind.del4$gene.names,cpm.lg.d4)
cpm.nord4.names <- cbind(gene.names=ind.del4$gene.names,cpm.nor.d4)
cpm.nlgd4.names <- cbind(gene.names=ind.del4$gene.names,cpm.nlg.d4)

y.wt<-DGEList(counts=ind.wt[,2:18],genes= ind.wt$gene.names)
y.wt<-calcNormFactors(y.wt) 
plotMDS(y.wt)
plotMDS(cpm(y.wt,log=TRUE))
cpm.un.wt <- cpm( y.wt, normalized.lib.sizes = FALSE, log = FALSE)
cpm.lg.wt <- cpm( y.wt, normalized.lib.sizes = FALSE, log = TRUE) 
cpm.nor.wt <- cpm(y.wt,normalized.lib.sizes = TRUE,log= FALSE)
cpm.nlg.wt <- cpm(y.wt,normalized.lib.sizes = TRUE,log= TRUE, prior.count=0.5)
boxplot(cpm.un.wt, outline= F)
boxplot(cpm.lg.wt, col="blue",outline= F)
boxplot(cpm.nor.wt, col="red",outline = F)
boxplot(cpm.nlg.wt, col="green",outline = F)
cpm.lgwt.names <- cbind(gene.names=ind.wt$gene.names,cpm.lg.wt)
cpm.norwt.names <- cbind(gene.names=ind.wt$gene.names,cpm.nor.wt)
cpm.nlgwt.names <- cbind(gene.names=ind.wt$gene.names,cpm.nlg.wt)
```
----
[,]means lines or rows of matrix ##DGEList() Creates a DGEList object from a table of counts (rows=features, columns=samples), group indicator for each column, library size (optional) and a table of feature annotation (optional).
TMM normalization, based on the assumption that the majority of genes are not DEGs; calcNormFactors() Calculate normalization factors to scale the raw library sizes.
plotMDS() plot samples on a two-dimensional scatterplot so that distances on the plot approximate the typical log2 fold changes between the samples.
cpm() computes counts per million; rpkm() computes counts per kilobase per million
----
# 4. create design matrix

```{r}
condition.del4 <- with(ind.sdel4,paste(Genotype,Timepoint,sep=".")) 
dsgn.del4 <- model.matrix(~0+condition.del4)
colnames(dsgn.del4) <- gsub("condition.del4","",colnames(dsgn.del4))
contrast.matrix.del4 <-makeContrasts(
  del4vsdel4.20min = del4.20m - del4.pre,
  del4vsdel4.30min = del4.30m - del4.pre,
  del4vsdel4.45min = del4.45m - del4.pre,
  del4vsdel4.60min = del4.60m - del4.pre,
  del4vsdel4.90min = del4.90m - del4.pre,
  del4vsdel4.120min = del4.120m - del4.pre,
  del4vsdel4.150min = del4.150m - del4.pre,
  del4vsdel4.180min = del4.180m - del4.pre,
  del4vsdel4.240min = del4.240m - del4.pre,
  levels=dsgn.del4)

condition.wt <- with(ind.swt,paste(Genotype,Timepoint,sep="."))
dsgn.wt <- model.matrix(~0+condition.wt)
colnames(dsgn.wt) <- gsub("condition.wt","",colnames(dsgn.wt))
contrast.matrix.wt <-makeContrasts(
  wtvswt.20min = wt.20m - wt.pre,
  wtvswt.30min = wt.30m - wt.pre,
  wtvswt.45min = wt.45m - wt.pre,
  wtvswt.60min = wt.60m - wt.pre,
  wtvswt.90min = wt.90m - wt.pre,
  wtvswt.120min = wt.120m - wt.pre,
  wtvswt.150min = wt.150m - wt.pre,
  wtvswt.180min = wt.180m - wt.pre,
  wtvswt.240min = wt.240m - wt.pre,
  levels=dsgn.wt)
```
----
read.table() reads a file in table format and creates a data frame from it
with(data,expr,...) data to use for constructing an environment
paste() concatenates vectors after converting to character; paste("my","hobby",sep=".")
sub() abd gsub() perform replacement of the fist and all matches respectively
??? model.matrix(~0+condition)
makeContrasts(..., contrasts=NULL, levels) 
??? what is the logic behind model.matrix and makeContrast matrix, why we need to set levels in contrast matrix
expressions, or character strings which can be parsed to expressions, specifying contrasts
contrasts=:	character vector specifying contrasts
levels character vector or factor giving the names of the parameters of which contrasts are desired, or a design matrix or other object with the parameter names as column names.
----
----
Question: when we want to compare more than two groups ( not use moderated t-test ), How to use different statistical method to find DEGs,

functions:
estimateDisp() 
Maximizes the negative binomial likelihood to give the estimate of the different dispersions across all tags.
Think abouT the relationship between normal distribution & Binomial distribution & Negative binomial distribution
Why they assume the dispersion of the data is negative binomial distribution? (variance is higher than mean, compare to Poisson)
estimateDisp()
glmQLFit()
glmQLFTest()
glmTreat()
topTags()
----

# 5.DEGs Analysis
## 5.0 GLM(edgeR)
```{r}
y <- estimateDisp(y, dsgn.del4, robust=TRUE)
y$common.dispersion
plotBCV(y)
fit<-glmQLFit(y,dsgn.del4,robust=TRUE)
head(fit$coefficients)
plotQLDisp(fit)
qlf<-glmQLFTest(fit,contrast=contrast.matrix)
topTags(qlf)
summary(decideTests(qlf))
plotMD(qlf)
anov<-glmTreat(fit,contrast=contrast.matrix)
topTags(anov)
write.table(topTags(anov,n=1000L, adjust.method="BH"),file="anov-wt-FC.txt",sep='\t',quote=F,row.names=F)
```

## 5.1. voom
```{r}
v.del4 <- voom (y.del4,dsgn.del4,plot=TRUE)
v.wt <- voom (y.wt,dsgn.wt,plot=TRUE)
### linear fit of deg data
fit.del4 <- lmFit(v.del4,dsgn.del4)
fit1 <- contrasts.fit(fit.del4,contrast.matrix.del4)
fit.del4 <- eBayes(fit1)

fit.wt <- lmFit(v.wt,dsgn.wt)
fit2 <- contrasts.fit(fit.wt,contrast.matrix.wt)
fit.wt <- eBayes(fit2)
### sreening diff probeid
dif<-topTable(fit.del4,coef=c("del4vsdel4.20min","del4vsdel4.30min","del4vsdel4.45min","del4vsdel4.60min","del4vsdel4.90min","del4vsdel4.120min","del4vsdel4.150min","del4vsdel4.180min","del4vsdel4.240min"),n=nrow(fit.del4),lfc=log2(4),adjust="BH")
dif2<-dif[dif[,"adj.P.Val"]<0.05,]
write.table(dif2,file="del4-DEGs.txt",sep='\t',quote=F,row.names=F)

dif3 <- topTable(fit.wt,coef=c("wtvswt.20min","wtvswt.30min","wtvswt.45min","wtvswt.60min","wtvswt.90min","wtvswt.120min","wtvswt.150min","wtvswt.180min","wtvswt.240min"),n=nrow(fit.wt),lfc=log2(4),adjust="BH")
dif4<-dif3[dif3[,"adj.P.Val"]<0.05,]
write.table(dif4,file="wt-DEGs.txt",sep='\t',quote=F,row.names=F)
### screening overall FC value of del4 groups
dif5<-topTable(fit.del4,coef=c("del4vsdel4.20min","del4vsdel4.30min","del4vsdel4.45min","del4vsdel4.60min","del4vsdel4.90min","del4vsdel4.120min","del4vsdel4.150min","del4vsdel4.180min","del4vsdel4.240min"),n=nrow(fit.del4),adjust="BH")
write.table(dif5,file="del4-all.txt",sep='\t',quote=F,row.names=F)
```
----
Given a linear model fit to microarray data, compute estimated coefficients and standard errors for a given set of contrasts.
contrasts.fit(fit, contrasts=NULL, coefficients=NULL)

How to compare the DEGs between wt and delpho4 strains under no-Pi conditions?


----


# 6. find DEGs dependent on Pho4 (indirect and direct targets)
```{r}
## get the max fold change value of DEGs of wt no-Pi treated groups
dif.wt<-read.table("wt-DEGs.txt",
                    sep='\t',header=T,row.names=1)
dif.wt<-dif.wt[,1:9]
fcw <- dif4[,1:10]
fcw[,"max"] <- apply(fcw[,2:10],1,max)
fcwt <- as.data.frame(cbind(fcw$genes,fcw$max));colnames(fcwt)<-c("genes","max")
write.table(fcwt,file="max-wt.txt",sep='\t',quote=F,row.names=F)
fc.wt <- read.table("max-wt.txt",sep='\t',header=T,row.names=1)
## filter out downregulated DEGs
fc.wt1 <- subset(fc.wt,max>=2,max)
## get the max fold change value of all genes del4 no-Pi treated groups
fcd <- dif5[,1:10]
fcd[,"max"] <- apply(fcd[,2:10],1,max)
fcd4 <- as.data.frame(cbind(fcd[,"genes"],fcd[,"max"]));colnames(fcd4)<-c("genes","max")
write.table(fcd4,file="max-del4.txt",sep='\t',quote=F,row.names=F)
fc.d4 <- read.table("max-del4.txt",sep='\t',header=T,row.names=1)
## find the DEGs that are pho4 target candidates
m1 <- match(rownames(fc.wt1),rownames(fc.d4))
m2 <- fc.d4[m1,]
m3 <- cbind(fc.wt1,m2); colnames(m3) <- c("wt.max","del4.max")
m4 <- transform(m3,minus=m3$wt.max-m3$del4.max)
m5 <- subset(m4,minus>1)
m55 <- subset(m4,minus>1.5)
m6 <- subset(m4,minus>2)
m7 <- subset(m4,minus>2.5)
m8 <- subset(m4,minus>3)
write.table(m5,file="pho4-cans.txt",sep='\t',quote=F)
write.table(m6,file="pho4-cans2.txt",sep='\t',quote=F)
```
----
## DEGs that have more than 8 times fold change between wt and del4 groups
myGenePlot(genes = c("CAGL0A01243g","CAGL0B02475g","CAGL0K07546g","CAGL0K07524g","CAGL0I04554g","CAGL0G06050g","CAGL0B02926g","CAGL0C00209g","CAGL0D04400g","CAGL0L09295g"))
----
# 7.get heatmap of foldchange in DEGs
```{r}
dif.wt<-read.table("wt-DEGs.txt",
                    sep='\t',header=T,row.names=1)
dif.wt<-dif.wt[,1:9]
pheatmap(dif.wt,
         color=colorRampPalette(c("blue","black","yellow"))(100),
         fontsize_row=0.5,
         fontsize_col=0.5,scale="none",
         border_color=NA,cluster_col = FALSE)

dif.del4 <-read.table("del4-DEGs.txt",
                    sep='\t',header=T,row.names=1)
dif.del4<-dif.del4[,1:9]
pheatmap(dif.del4,
         color=colorRampPalette(c("blue","black","yellow"))(100),
         fontsize_row=0.5,
         fontsize_col=0.5,scale="none",
         border_color=NA,cluster_col = FALSE)
```
## 7.1 use heatmap to visualize genes with lower foldchange
```{r}
thres.wt <- dif.wt>=6
dif.wt[thres.wt]=6
pheatmap(dif.wt,
         color=colorRampPalette(c("blue","black","yellow"))(100),
         fontsize_row=0.5,
         fontsize_col=0.5,scale="none",
         border_color=NA,cluster_col = FALSE)
thres.del4 <- dif.del4>=5
dif.del4[thres.del4]=5
pheatmap(dif.del4,
         color=colorRampPalette(c("blue","black","yellow"))(100),
         fontsize_row=0.5,
         fontsize_col=0.5,scale="none",
         border_color=NA,cluster_col = FALSE)

```
